internal/kafka/internal/services/kafkatlscertmgmt/secure_storage.go

```diff
@@ -48,18 +51,41 @@ func newSecureStorage(config *config.AWSConfig, automaticCertificateManagementCo
 		return nil, err
 	}
 	return &secureStorage{
-		secretClient: secretClient,
-		secretCache:  secretCache,
-		secretPrefix: config.SecretManager.SecretPrefix,
+		secretClient:     secretClient,
+		secretCache:      secretCache,
+		lock:             sync.Mutex{},
+		storageItemLocks: map[string]*sync.Mutex{},
+		secretPrefix:     config.SecretManager.SecretPrefix,
 	}, nil
 }
 
+// Lock blocks until it can acquire the name-specific lock.
+// The process of obtaining a name-specific lock is atomic, as required per the certmagic.Locker
+// interface documentation: https://pkg.go.dev/github.com/caddyserver/certmagic#Locker.
+// NOTE: The intent here is to not implement a distributed locking mechanism but an in-process locking mechanism good enough for
+// our certificate management logic which happens within one reconciliater leader.
 func (storage *secureStorage) Lock(ctx context.Context, key string) error {
-	return nil // NOOP as AWS Secret manager uses versioning mechanism to achieve optimistic locking
+	storage.lock.Lock()
+	mu, ok := storage.storageItemLocks[key]
+	if !ok {
+		// initialize the storage item lock the first time if it does not exist yet
+		mu = &sync.Mutex{}
+		storage.storageItemLocks[key] = mu
+	}
+	storage.lock.Unlock()
```

Comment:
This can be optimised to avoid locking when not needed (double-checked locking pattern):
```suggestion
	mu, ok := storage.storageItemLocks[key]
	if !ok {
		storage.lock.Lock()
		mu, ok = storage.storageItemLocks[key]
		if !ok {
			// initialize the storage item lock the first time if it does not exist yet
			mu = &sync.Mutex{}
			storage.storageItemLocks[key] = mu
		}
		storage.lock.Unlock()
	}
```

Reply:
Thanks. Yes, I considered this initially and decided to went with the current approach in the PR because the operation that happens in between the `storage.lock.Lock()` and `storage.lock.Unlock()` is not that time consuming. 
Definitely this pattern is useful when acquiring a key item lock becomes an expensive operation e.g when we want to implement a distributed lock